tests/test_cliCmdQueryDataIds.py

```diff
@@ -0,0 +1,220 @@
+from lsst.daf.butler.tests import MetricsExample
+
+
+TESTDIR = os.path.abspath(os.path.dirname(__file__))
+
+
+class QueryDataIdsTest(unittest.TestCase):
+
+    configFile = os.path.join(TESTDIR, "config/basic/butler.yaml")
+    storageClassFactory = StorageClassFactory()
+
+    @staticmethod
+    def _makeExampleMetrics():
+        return MetricsExample({"AM1": 5.2, "AM2": 30.6},
+                              {"a": [1, 2, 3],
+                               "b": {"blue": 5, "red": "green"}},
+                              [563, 234, 456.7, 752, 8, 9, 27])
+
+    @staticmethod
+    def _addDatasetType(datasetTypeName, dimensions, storageClass, registry):
+        """Create a DatasetType and register it
+        """
+        datasetType = DatasetType(datasetTypeName, dimensions, storageClass)
+        registry.registerDatasetType(datasetType)
+        return datasetType
+
+    @staticmethod
+    def _queryDataIds(repo, dimensions=(), collections=(), datasets=None, where=None, ):
+        return script.queryDataIds(repo=repo,
+                                   dimensions=dimensions,
+                                   collections=collections,
+                                   datasets=datasets,
+                                   where=where)
+
+    def _assertTablesEqual(self, tables, expectedTables):
+        """Verify that a list of astropy tables matches a list of expected
+        astropy tables."""
+        diff = io.StringIO()
+        self.assertEqual(len(tables), len(expectedTables))
+        for table, expected in zip(tables, expectedTables):
+            self.assertTrue(report_diff_values(table, expected, fileobj=diff), msg=diff.getvalue())
+
+    def setUp(self):
+        self.root = tempfile.mkdtemp(dir=TESTDIR)
+        Butler.makeRepo(self.root, config=Config(self.configFile))
+        self.butlerConfigFile = os.path.join(self.root, "butler.yaml")
+        self.storageClassFactory.addFromConfig(self.configFile)
+
+        # New datasets will be added to run and tag, but we will only look in
+        # tag when looking up datasets.
+        run = "ingest/run"
+        tag = "ingest"
+        self.butler = Butler(self.butlerConfigFile, run=run, collections=[tag], tags=[tag])
+
+        # There will not be a collection yet
+        collections = set(self.butler.registry.queryCollections())
+        self.assertEqual(collections, set([run, tag]))
+
+        storageClass = self.storageClassFactory.getStorageClass("StructuredCompositeReadComp")
+
+        # Create and register a DatasetType
+        dimensions = self.butler.registry.dimensions.extract(["instrument", "visit"])
+        datasetTypeName = "test_metric_comp"
+        self.datasetType = self._addDatasetType(datasetTypeName, dimensions, storageClass,
+                                                self.butler.registry)
+
+        # Add needed Dimensions
+        self.butler.registry.insertDimensionData("instrument", {"name": "DummyCamComp"})
```

Comment:
I can't work out how to make it do what I need, and I think it mocks some behaviors in a way that's hard to understand (it mentions a registry override, and arbitrary cross assignment of relationships). In a day I couldn't get it to do what I needed, Jim directed me at something else (test_simpleButler), and I understand what this code does (and I think it's clear from reading it, but maybe just because it's made of stuff that I do grok). So, I'd like to leave it as is if that's ok.

Reply:
Ok. I'm not sure what the real problem is because makeTestRepo doesn't mock anything -- you can call it with empty dataIds iterables and it's still less code than calling Butler.makeRepo yourself isn't it?